src/common/summarize-profile.js

```diff
@@ -108,23 +94,51 @@ export function summarizeProfile(profile: Profile) {
  * are cached between calls.
  * @returns {function} Function categorizer.
  */
-function functionNameCategorizer() {
-  const cache = new Map();
-  return function functionNameToCategory(name) {
-    const existingCategory = cache.get(name);
-    if (existingCategory !== undefined) {
+function functionNameCategorizer(numFunctions): (number, number[], string[]) => string {
+  const cache = new Array(numFunctions);
+  cache.fill('', 0, numFunctions);
```

Comment:
These are both for the same reason, and if the indexes are nonsequential then that reason is dead.

This all comes from me spending too much time with JITs. You are correct that this will use more memory. The only reason to do it is to make this monomorphic -- so that every lookup from 'cache' and everything that uses categories can be type-specialized to a string. In theory, this will make the JITted code faster. But it's totally possible that either the code type-specialized for string-or-undefined will be just as fast, or that the extra fill() will actually slow things down.

If I could figure out where my timing numbers are getting printed, I would test it out. If the difference is negligible (which is likely), then I wouldn't bother with it.